### PR TITLE
Refactored ReorderLiteralsTransformer

### PR DESCRIPTION
**Pull request by azreika**
Refactored the ReorderLiteralsTransformer. Main goal was to soon get it ready to use in the Ast2Ram translator.

Main changes:
* Moved BindingStore out of the MST and into its own class
    * This is a much more accurate and easier-to-use measure of dynamic variable-bindings originally created for the MST
* Extended the BindingStore into a more useful black-box interface for binding-checks
* Updated the reorderer + SIPS functions to use the new binding store mechanism
* Moved reordering functionality into public static methods in the transformer - necessary for use later on

### CODE REVIEW

File: src/ast/BindingStore.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
+/*
```
Comment (mmcgr) [inline src/ast/BindingStore.h:1]: Maybe we should have `ast/utility/` for files/classes like this?

File: src/ast/BindingStore.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    /** Check if a variable is bound */
+    bool isBound(std::string varName) const {
+        return contains(stronglyBoundVariables, varName) || contains(weaklyBoundVariables, varName);
+    }
+
+    /** Check if an argument is bound */
+    bool isBound(const AstArgument* arg) const;
+
+    /** Counts the number of bound arguments in the given atom */
+    unsigned int numBoundArguments(const AstAtom* atom) const;
```
Comment (mmcgr) [inline src/ast/BindingStore.h:65]: `size_t` would be more consistent than `unsigned int` for the size/count of things.

File: src/ast/BindingStore.h (hunk 1 above)
Comment (azreika) [reply src/ast/BindingStore.h:1]: I agree! We should add `ast/utility/` and throw in AstUtils and BindingStore into there. Do you want me to do it as part of this PR?

Comment (mmcgr) [reply src/ast/BindingStore.h:1]: I think separate PR so we can think about it a bit more. There are other files that could also move.

**Review by mmcgr — APPROVED** [review]

**Pull request merged**